File: static/js/purchase_transactions.js
// File: static/js/purchase_transactions.js

document.addEventListener('DOMContentLoaded', function() {
    const newTransactionForm = document.getElementById('new-transaction-form');
    const editTransactionForm = document.getElementById('edit-transaction-form');
    const transactionsTable = document.getElementById('transactions-table');
    const modal = document.getElementById('transaction-modal');
    const vendorSelect = document.getElementById('vendor');
    const productSelect = document.getElementById('product');
    const quantityInput = document.getElementById('quantity');
    const unitPriceInput = document.getElementById('unit-price');
    const totalAmountInput = document.getElementById('total-amount');

    vendorSelect.addEventListener('change', loadVendorProducts);
    productSelect.addEventListener('change', updateUnitPrice);
    quantityInput.addEventListener('input', updateTotalAmount);
    unitPriceInput.addEventListener('input', updateTotalAmount);
    newTransactionForm.addEventListener('submit', addTransaction);
    editTransactionForm.addEventListener('submit', updateTransaction);
    transactionsTable.addEventListener('click', handleTableActions);

    function loadVendorProducts() {
        const vendorId = vendorSelect.value;
        productSelect.innerHTML = '<option value="">Select a product</option>';
        productSelect.disabled = true;
        
        if (vendorId) {
            fetch(`/vendor-products/${vendorId}`)
                .then(response => response.json())
                .then(products => {
                    products.forEach(product => {
                        const option = document.createElement('option');
                        option.value = product.id;
                        option.textContent = `${product.name} (${product.sku})`;
                        option.dataset.price = product.default_price;
                        productSelect.appendChild(option);
                    });
                    productSelect.disabled = false;
                });
        }
    }

    function updateUnitPrice() {
        const selectedOption = productSelect.options[productSelect.selectedIndex];
        if (selectedOption.dataset.price) {
            unitPriceInput.value = parseFloat(selectedOption.dataset.price).toFixed(2);
            updateTotalAmount();
        }
    }

    function updateTotalAmount() {
        const quantity = parseFloat(quantityInput.value) || 0;
        const unitPrice = parseFloat(unitPriceInput.value) || 0;
        totalAmountInput.value = (quantity * unitPrice).toFixed(2);
    }

    function addTransaction(e) {
        e.preventDefault();
        const formData = {
            vendor_id: vendorSelect.value,
            date: document.getElementById('date').value,
            product_id: productSelect.value,
            quantity: quantityInput.value,
            unit_price: unitPriceInput.value
        };

        fetch('/purchase-transactions', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData),
        })
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                location.reload();  // Refresh to show new transaction
            }
        })
        .catch((error) => {
            console.error('Error:', error);
        });
    }

    function handleTableActions(e) {
        if (e.target.classList.contains('edit-btn')) {
            const row = e.target.closest('tr');
            const id = row.dataset.id;
            editTransaction(id);
        } else if (e.target.classList.contains('delete-btn')) {
            const row = e.target.closest('tr');
            const id = row.dataset.id;
            deleteTransaction(id);
        }
    }

    function editTransaction(id) {
        fetch(`/purchase-transactions/${id}`)
            .then(response => response.json())
            .then(data => {
                document.getElementById('edit-transaction-id').value = data.id;
                document.getElementById('edit-vendor').value = data.vendor_id;
                document.getElementById('edit-date').value = data.date;
                loadVendorProductsForEdit(data.vendor_id, data.product_id);
                document.getElementById('edit-quantity').value = data.quantity;
                document.getElementById('edit-unit-price').value = data.unit_price.replace(/[^0-9.]/g, '');
                document.getElementById('edit-total-amount').value = data.total_amount;
                modal.style.display = 'block';
            });
    }

    function loadVendorProductsForEdit(vendorId, productId) {
        const editProductSelect = document.getElementById('edit-product');
        editProductSelect.innerHTML = '<option value="">Select a product</option>';
        
        fetch(`/vendor-products/${vendorId}`)
            .then(response => response.json())
            .then(products => {
                products.forEach(product => {
                    const option = document.createElement('option');
                    option.value = product.id;
                    option.textContent = `${product.name} (${product.sku})`;
                    option.dataset.price = product.default_price;
                    editProductSelect.appendChild(option);
                });
                editProductSelect.value = productId;
            });
    }

    function updateTransaction(e) {
        e.preventDefault();
        const id = document.getElementById('edit-transaction-id').value;
        const formData = {
            vendor_id: document.getElementById('edit-vendor').value,
            date: document.getElementById('edit-date').value,
            product_id: document.getElementById('edit-product').value,
            quantity: document.getElementById('edit-quantity').value,
            unit_price: document.getElementById('edit-unit-price').value
        };

        fetch(`/purchase-transactions/${id}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData),
        })
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                modal.style.display = 'none';
                location.reload();  // Refresh to show updated transaction
            }
        })
        .catch((error) =>

File: views/purchase_transactions.py
# File: views/purchase_transactions.py

from flask import render_template, request, jsonify
from models import PurchaseTransaction, Vendor, Product, Inventory
from utils import DateUtil, CurrencyUtil
from datetime import datetime

def purchase_transactions():
    if request.method == 'GET':
        transactions = PurchaseTransaction.query.all()
        vendors = Vendor.query.all()
        products = Product.query.all()
        return render_template('purchase_transactions.html', 
                               transactions=transactions, 
                               vendors=vendors,
                               products=products)
    elif request.method == 'POST':
        data = request.json
        new_transaction = PurchaseTransaction(
            vendor_id=data['vendor_id'],
            date=datetime.strptime(data['date'], '%Y-%m-%d'),
            product_id=data['product_id'],
            quantity=data['quantity'],
            unit_price=data['unit_price'],
            total_amount=data['quantity'] * data['unit_price']
        )
        db.session.add(new_transaction)
        
        # Update inventory
        inventory = Inventory.query.filter_by(product_id=data['product_id']).first()
        if inventory:
            inventory.quantity += data['quantity']
        else:
            new_inventory = Inventory(product_id=data['product_id'], quantity=data['quantity'])
            db.session.add(new_inventory)
        
        db.session.commit()
        return jsonify({'success': True, 'id': new_transaction.id})

def get_transaction(transaction_id):
    transaction = PurchaseTransaction.query.get_or_404(transaction_id)
    return jsonify({
        'id': transaction.id,
        'vendor': transaction.vendor.name,
        'date': DateUtil.format_date(transaction.date),
        'product': transaction.product.name,
        'quantity': transaction.quantity,
        'unit_price': CurrencyUtil.format_currency(transaction.unit_price),
        'total_amount': CurrencyUtil.format_currency(transaction.total_amount)
    })

def update_transaction(transaction_id):
    transaction = PurchaseTransaction.query.get_or_404(transaction_id)
    data = request.json
    
    # Calculate the change in quantity
    quantity_change = data['quantity'] - transaction.quantity
    
    transaction.vendor_id = data['vendor_id']
    transaction.date = datetime.strptime(data['date'], '%Y-%m-%d')
    transaction.product_id = data['product_id']
    transaction.quantity = data['quantity']
    transaction.unit_price = data['unit_price']
    transaction.total_amount = data['quantity'] * data['unit_price']
    
    # Update inventory
    inventory = Inventory.query.filter_by(product_id=data['product_id']).first()
    if inventory:
        inventory.quantity += quantity_change
    else:
        new_inventory = Inventory(product_id=data['product_id'], quantity=quantity_change)
        db.session.add(new_inventory)
    
    db.session.commit()
    return jsonify({'success': True})

def delete_transaction(transaction_id):
    transaction = PurchaseTransaction.query.get_or_404(transaction_id)
    
    # Update inventory
    inventory = Inventory.query.filter_by(product_id=transaction.product_id).first()
    if inventory:
        inventory.quantity -= transaction.quantity
        if inventory.quantity < 0:
            inventory.quantity = 0
    
    db.session.delete(transaction)
    db.session.commit()
    return jsonify({'success': True})

def get_vendor_products(vendor_id):
    vendor = Vendor.query.get_or_404(vendor_id)
    products = Product.query.filter_by(vendor_id=vendor_id).all()
    return jsonify([{
        'id': p.id,
        'name': p.name,
        'sku': p.sku,
        'default_price': CurrencyUtil.format_currency(p.default_price)
    } for p in products])